src/core/reducers/data/personalization.ts

```diff
@@ -0,0 +1,73 @@
+import { createTransform, persistReducer } from 'redux-persist';
+import storage from 'redux-persist/lib/storage';
+import Actions from '../../actions';
+import Adapter from '../../adapters/personalization';
+import Selectors from '../../selectors';
+import Store from '../../store';
+
+export type Action = Actions.UpdateBiasing;
+export type State = Store.Personalization;
+
+export const DEFAULT: State = {
+  biasing: {
+    allIds: [],
+    byId: {},
+    expiry: 0
+  }
+};
+
+function updatePersonalization(state: State = DEFAULT, action: Action): State {
+  switch (action.type) {
+    case Actions.UPDATE_BIASING: return updateBiasing(state, action.payload);
+    default: return state;
+  }
+}
+
+export const updateBiasing = (state: State, payload: Actions.Payload.Personalization.Biasing) => {
+  const byId = {
+    ...state.biasing.byId,
+    [payload.variant]: {
+      ...state.biasing.byId[payload.variant],
+      [payload.key]: payload.bias
+    }
+  };
+
+  let allIds = insertSorted(state.biasing.allIds, { variant: payload.variant, key: payload.key });
+  const keyCount = Object.keys(byId[payload.variant]).length;
+  const config = payload.config;
+  if (config.attributes[payload.variant] && (keyCount > config.attributes[payload.variant].maxBiases)) {
+    allIds = removeLast(allIds, payload.variant);
+  }
+  if (allIds.length > config.maxBiases) {
+    allIds = allIds.slice(0, config.maxBiases);
+  }
+
+  return { ...state, biasing: { ...state.biasing, byId, allIds }};
+};
+
+// tslint:disable-next-line max-line-length
+export const insertSorted = (allIds: Store.Personalization.BiasKey[], { variant, key }: Store.Personalization.BiasKey) => {
+  const noDuplicate = allIds.filter((id) => !(id.variant === variant && id.key === key));
+  return [{ variant, key }, ...noDuplicate]; //TODO: insert sorted
+};
+
+export const removeLast = (allIds, variant) => {
```

Comment:
More appropriate name ```removeVariant```

Reply:
How about something like `removeLastForVariant`. I agree that `removeLast` is not a good name, but what the function does is it finds the item for that variant furthest back in the array and removes it. It doesn't remove all items of that variant.

Reply:
Oh I didn't even notice the for loop was going backwards lol.

Reply:
Maybe removeOldest?